lib/WasmReader/BaseWasmReader.h

```diff
@@ -8,19 +8,36 @@
 namespace Wasm
 {
 
-class BaseWasmReader
-{
-public:
-    virtual WasmOp ReadFromModule() = 0;
-    virtual WasmOp ReadExpr() = 0;
-    virtual WasmOp ReadFromBlock() = 0;
-    virtual WasmOp ReadFromCall() = 0;
-    virtual bool IsBinaryReader() = 0;
-    WasmNode    m_currentNode;
-    ModuleInfo * m_moduleInfo;
-
-protected:
-    WasmFunctionInfo *  m_funcInfo;
-};
+    enum SectionFlag
+    {
+        fSectNone,
+        fSectIgnore,
+    };
+
+#define WASM_SECTION(name, id, flag) bSect ## name,
+    enum SectionCode
+    {
+        bSectInvalid = -1,
+    #include "WasmSections.h"
+        bSectLimit
+    };
+
+    class BaseWasmReader
+    {
+    public:
+        virtual void InitializeReader() = 0;
+        virtual bool ReadNextSection(SectionCode nextSection) = 0;
+        virtual bool ProcessSection(SectionCode nextSection) = 0;
```

Comment:
ProcessSection means the reader will take care of doing the work for that section. Still need to come up with a nice way to return back to the bytecode generator if needed.